skimage/feature/sift.py

```diff
@@ -0,0 +1,729 @@
+        >>> detector_extractor2 = SIFT()
+        >>> detector_extractor1.detect_and_extract(img1)
+        >>> detector_extractor2.detect_and_extract(img2)
+        >>> matches = match_descriptors(detector_extractor1.descriptors,
+        ...                             detector_extractor2.descriptors,
+        ...                             max_ratio=0.6)
+        >>> matches[10:15]
+        array([[ 10,  71],
+               [ 11,  65],
+               [ 12,  66],
+               [ 13, 369],
+               [ 15, 366]])
+        >>> detector_extractor1.keypoints[matches[10:15, 0]]
+        array([[126, 417],
+               [126, 423],
+               [126, 423],
+               [127, 188],
+               [132, 190]])
+        >>> detector_extractor2.keypoints[matches[10:15, 1]]
+        array([[ 94, 126],
+               [ 88, 126],
+               [ 88, 126],
+               [323, 127],
+               [321, 132]])
+
+        """
+
+    def __init__(self, upsampling=2, n_octaves=8, n_scales=3, sigma_min=1.6,
+                 sigma_in=0.5,
+                 c_dog=0.04 / 3, c_edge=10, n_bins=36, lambda_ori=1.5,
+                 c_max=0.8, lambda_descr=6, n_hist=4, n_ori=8):
+        if upsampling in [1, 2, 4]:
+            self.upsampling = upsampling
+        else:
+            raise ValueError("upsampling must be 1, 2 or 4")
+        self.n_octaves = n_octaves
+        self.n_scales = n_scales
+        self.sigma_min = sigma_min / upsampling
+        self.sigma_in = sigma_in
+        self.c_dog = (2 ** (1 / n_scales) - 1) / (2 ** (1 / 3) - 1) * c_dog
+        self.c_edge = c_edge
+        self.n_bins = n_bins
+        self.lambda_ori = lambda_ori
+        self.c_max = c_max
+        self.lambda_descr = lambda_descr
+        self.n_hist = n_hist
+        self.n_ori = n_ori
+        self.delta_min = 1 / upsampling
+        self.deltas = self._deltas()
+        self.scalespace_sigmas = None
+        self.keypoints = None
+        self.positions = None
+        self.sigmas = None
+        self.scales = None
+        self.orientations = None
+        self.octaves = None
+        self.descriptors = None
+
+    def _set_number_of_octaves(self, image_shape):
+        size_min = 12  # minimum size of last octave
+        s0 = min(image_shape) * self.upsampling
+        max_octaves = int(math.log2(s0 / size_min) + 1)
+        if max_octaves < self.n_octaves:
+            self.n_octaves = max_octaves
+
+    def _deltas(self, dtype=float):
+        deltas = self.delta_min * np.power(2, np.arange(self.n_octaves))
+        return deltas.astype(dtype, copy=False)
+
+    def _create_scalespace(self, image):
+        """Source: "Anatomy of the SIFT Method" Alg. 1
+        Construction of the scalespace by gradually blurring (scales) and
+        downscaling (octaves) the image.
+        """
+        scalespace = []
+        dtype = image.dtype
+        if self.upsampling > 1:
+            image = _oversample_bilin(image, self.deltas[0])
```

Comment:
```suggestion
        if self.upsampling > 1:
            image = np.ascontiguousarray(image)        
            image = _oversample_bilin(image, self.deltas[0])
```
Ensure the input to the Cython function is C-contiguous as expected.

Reply:
Sorry, forgot that.